load-balancer/power-of-two-choices/src/main/java/io/smallrye/stork/loadbalancer/poweroftwochoices/PowerOfTwoChoicesLoadBalancer.java

```diff
@@ -0,0 +1,51 @@
+package io.smallrye.stork.loadbalancer.poweroftwochoices;
+
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Random;
+
+import io.smallrye.stork.api.LoadBalancer;
+import io.smallrye.stork.api.NoServiceInstanceFoundException;
+import io.smallrye.stork.api.ServiceInstance;
+import io.smallrye.stork.impl.ServiceInstanceWithStatGathering;
+import io.smallrye.stork.loadbalancer.requests.InflightRequestCollector;
+
+/**
+ * Select two random destinations and then select the one with the least assigned requests. This avoids the overhead of
+ * least-requests and the worst case for random where it selects a busy destination.
+ */
+public class PowerOfTwoChoicesLoadBalancer implements LoadBalancer {
+
+    private final InflightRequestCollector collector = new InflightRequestCollector();
+    private final Random random = new SecureRandom();
+
+    @Override
+    public ServiceInstance selectServiceInstance(Collection<ServiceInstance> serviceInstances) {
+        if (serviceInstances.isEmpty()) {
+            throw new NoServiceInstanceFoundException("No service instance found");
+        }
+
+        List<ServiceInstance> instances = new ArrayList<>(serviceInstances);
+        int count = instances.size();
+
+        if (count == 1) {
+            return instances.get(0);
+        }
+
+        ServiceInstance first = instances.get(random.nextInt(count));
+        ServiceInstance second = instances.get(random.nextInt(count));
```

Comment:
shouldn't it select a different random instance?

Reply:
actually, we select these random two every time so the fact that we sometimes have the same two is not an issue

Reply:
yes, it's not a problem.
I was wondering to do an object equality check in this case, but it's clearly premature optimization.